src/Build/Evaluation/Expander.cs

```diff
@@ -3268,6 +3275,122 @@ internal object Execute(object objectInstance, IPropertyProvider<T> properties,
                 }
             }
 
+            /// <summary>
+            /// Shortcut to avoid calling into binding if we recognize some most common functions.
+            /// Binding is expensive and throws first-chance MissingMethodExceptions, which is
+            /// bad for debugging experience and has a performance cost.
+            /// A typical binding operation with exception can take ~1.500 ms; this call is ~0.050 ms
+            /// (rough numbers just for comparison).
+            /// See https://github.com/Microsoft/msbuild/issues/2217
+            /// </summary>
+            /// <param name="objectInstance">Object that the function is called on</param>
+            /// <param name="args">arguments</param>
+            /// <returns>The value returned from the function call</returns>
+            private object ExecuteWellKnownFunction(object objectInstance, object[] args)
+            {
+                if (objectInstance is string text)
+                {
+                    if (string.Equals(_methodMethodName, "Substring", StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (args.Length == 1)
+                        {
+                            if (args[0] is string arg0 &&
+                                int.TryParse(arg0, out int startIndex) &&
```

Comment:
I don't know if it matters that we do all of this checking here.  If we call the method with bad args, we might as well let it throw.  Otherwise we'll probably pay the penalty twice since the method is checking the same stuff right?

Reply:
Right, that's what I was thinking too.